### PR TITLE
new resource "azurerm_synapse_private_link_hub"

### LINKED ISSUE #12493: Support for Synapse Private Link Hubs
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Add support for provisioning [Synapse Private Link Hubs](https://docs.microsoft.com/en-us/azure/synapse-analytics/security/synapse-private-link-hubs)

### New or Affected Resource(s)

* azurerm_synapse_private_link_hub

### Potential Terraform Configuration

```hcl
resource "azurerm_synapse_private_link_hub" "plhub" {
  name                = "myplhub"
  resource_group_name = "my-resource-group"
  location            = "West Europe"

  tags = {
    ENV = "Test"
  }
}

data "azurerm_synapse_private_link_hub" "example" {
  name                = "existing"
  resource_group_name = "example-resource-group"
}
```

### PR DESCRIPTION
**Pull request by owenfarrell**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #12493

Output from acceptance testing:

```
$ make acctests SERVICE='synapse' TESTARGS='-run=TestAccSynapsePrivateLinkHub_' TESTTIMEOUT='5m' 
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./azurerm/internal/services/synapse -run=TestAccSynapsePrivateLinkHub_ -timeout 5m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccSynapsePrivateLinkHub_create
=== PAUSE TestAccSynapsePrivateLinkHub_create
=== RUN   TestAccSynapsePrivateLinkHub_update
=== PAUSE TestAccSynapsePrivateLinkHub_update
=== CONT  TestAccSynapsePrivateLinkHub_create
=== CONT  TestAccSynapsePrivateLinkHub_update
--- PASS: TestAccSynapsePrivateLinkHub_create (171.43s)
--- PASS: TestAccSynapsePrivateLinkHub_update (249.49s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/synapse     252.111s
```

### CODE REVIEW

File: azurerm/internal/services/synapse/synapse_private_link_hub.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+func resourceSynapsePrivateLinkHubCreate(d *pluginsdk.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Synapse.PrivateLinkHubsClient
+	ctx, cancel := timeouts.ForCreate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	name := d.Get("name").(string)
+	resourceGroup := d.Get("resource_group_name").(string)
+
+	existing, err := client.Get(ctx, resourceGroup, name)
```
Comment (manicminer) [inline azurerm/internal/services/synapse/synapse_private_link_hub.go:66]: We generally prefer to generate a resource ID first and use it through the Create function

```suggestion
	subscriptionId := meta.(*clients.Client).Account.SubscriptionId
	id := parse.NewPrivateLinkHubID(subscriptionId, d.Get("resource_group_name").(string), d.Get("name").(string))

	existing, err := client.Get(ctx, id.ResourceGroup, id.Name)
```

File: azurerm/internal/services/synapse/synapse_private_link_hub.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+	if err != nil {
+		if !utils.ResponseWasNotFound(existing.Response) {
+			return fmt.Errorf("checking for present of existing Synapse Private Link Hub %q (Resource Group %q): %+v", name, resourceGroup, err)
```
Comment (manicminer) [inline azurerm/internal/services/synapse/synapse_private_link_hub.go:69]: The id struct has a useful String method which can be used for simpler and more consistent error messages

```suggestion
			return fmt.Errorf("checking for presence of %s: %+v", id, err)
```

File: azurerm/internal/services/synapse/synapse_private_link_hub.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+		}
+	}
+	if existing.ID != nil && *existing.ID != "" {
+		return tf.ImportAsExistsError("azurerm_synapse_private_link_hub", *existing.ID)
```
Comment (manicminer) [inline azurerm/internal/services/synapse/synapse_private_link_hub.go:73]: Using the generated id here will help avoid issues due to malformed API responses

```suggestion
	} else {
		return tf.ImportAsExistsError("azurerm_synapse_private_link_hub", id.ID())
```

File: azurerm/internal/services/synapse/synapse_private_link_hub.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+	}
+
+	privateLinkHubInfo := synapse.PrivateLinkHub{
+		Location: utils.String(location.Normalize(d.Get("location").(string))),
+		Tags:     tags.Expand(d.Get("tags").(map[string]interface{})),
+	}
+
+	_, err = client.CreateOrUpdate(ctx, privateLinkHubInfo, resourceGroup, name)
```
Comment (manicminer) [inline azurerm/internal/services/synapse/synapse_private_link_hub.go:81]: By generating the id first, we can use it everywhere we need its fields

```suggestion
	_, err = client.CreateOrUpdate(ctx, privateLinkHubInfo, id.ResourceGroup, id.Name)
```

File: azurerm/internal/services/synapse/synapse_private_link_hub.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+	if err != nil {
+		return fmt.Errorf("creating Synapse Private Link Hub %q (Resource Group %q): %+v", name, resourceGroup, err)
```
Comment (manicminer) [inline azurerm/internal/services/synapse/synapse_private_link_hub.go:83]: As above, simpler error construction

```suggestion
		return fmt.Errorf("creating %s: %+v", id, err)
```

File: azurerm/internal/services/synapse/synapse_private_link_hub.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+	}
+
+	subscriptionId := meta.(*clients.Client).Account.SubscriptionId
+	id := parse.NewPrivateLinkHubID(subscriptionId, resourceGroup, name)
```
Comment (manicminer) [inline azurerm/internal/services/synapse/synapse_private_link_hub.go:87]: (moved further up)
```suggestion
```

File: azurerm/internal/services/synapse/synapse_private_link_hub.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+	}
+
+	resp, err := client.Get(ctx, id.ResourceGroup, id.Name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[INFO] synapse %q does not exist - removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("retrieving Synapse Private Link Hub %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
```
Comment (manicminer) [inline azurerm/internal/services/synapse/synapse_private_link_hub.go:110]: Also here
```suggestion
		return fmt.Errorf("retrieving %s: %+v", id, err)
```

File: azurerm/internal/services/synapse/synapse_private_link_hub.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+	client := meta.(*clients.Client).Synapse.PrivateLinkHubsClient
+	ctx, cancel := timeouts.ForUpdate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.PrivateLinkHubID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	if d.HasChanges("tags") {
```
Comment (manicminer) [inline azurerm/internal/services/synapse/synapse_private_link_hub.go:130]: Can use d.HasChange if we're only checking one property
```suggestion
	if d.HasChange("tags") {
```

File: azurerm/internal/services/synapse/synapse_private_link_hub.go [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+		privateLinkHubPatchInfo := synapse.PrivateLinkHubPatchInfo{
+			Tags: tags.Expand(d.Get("tags").(map[string]interface{})),
+		}
+
+		_, err := client.Update(ctx, privateLinkHubPatchInfo, id.ResourceGroup, id.Name)
+		if err != nil {
+			return fmt.Errorf("updating Synapse Private Link Hub %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
```
Comment (manicminer) [inline azurerm/internal/services/synapse/synapse_private_link_hub.go:137]:
```suggestion
			return fmt.Errorf("updating %s: %+v", id, err)
```

File: azurerm/internal/services/synapse/synapse_private_link_hub.go [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+	defer cancel()
+
+	id, err := parse.PrivateLinkHubID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	future, err := client.Delete(ctx, id.ResourceGroup, id.Name)
+	if err != nil {
+		return fmt.Errorf("deleting Synapse Private Link Hub %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
```
Comment (manicminer) [inline azurerm/internal/services/synapse/synapse_private_link_hub.go:156]:
```suggestion
		return fmt.Errorf("deleting %s: %+v", id, err)
```

File: azurerm/internal/services/synapse/synapse_private_link_hub.go [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+	}
+
+	// sometimes the waitForCompletion rest api will return 404
+	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		if !response.WasNotFound(future.Response()) {
+			return fmt.Errorf("waiting for Synapse Private Link Hub %q (Resource Group %q) to be deleted: %+v", id.Name, id.ResourceGroup, err)
```
Comment (manicminer) [inline azurerm/internal/services/synapse/synapse_private_link_hub.go:162]:
```suggestion
			return fmt.Errorf("waiting for deletion of %s: %+v", id, err)
```

File: azurerm/internal/services/synapse/synapse_private_link_hub_test.go [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/acceptance/check"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/synapse/parse"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tf/pluginsdk"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+type SynapsePrivateLinkHubResource struct{}
+
+func TestAccSynapsePrivateLinkHub_create(t *testing.T) {
```
Comment (manicminer) [inline azurerm/internal/services/synapse/synapse_private_link_hub_test.go:18]: We prefer to align the test name and the config method name where there's a 1:1 correspondence

```suggestion
func TestAccSynapsePrivateLinkHub_basic(t *testing.T) {
```

File: azurerm/internal/services/synapse/synapse_private_link_hub_test.go [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+	return utils.Bool(true), nil
+}
+
+func (r SynapsePrivateLinkHubResource) basic(data acceptance.TestData) string {
+	template := r.template(data)
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_synapse_private_link_hub" "test" {
+  name                = "acctestsw%d"
```
Comment (manicminer) [inline azurerm/internal/services/synapse/synapse_private_link_hub_test.go:77]: Using indexed verbs will make it easier to read and reduce repetition if config needs to be added at a later date
```suggestion
%[1]s

resource "azurerm_synapse_private_link_hub" "test" {
  name                = "acctestsw%[2]d"
```

File: azurerm/internal/services/synapse/synapse_private_link_hub_test.go [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+`, template, data.RandomInteger)
+}
+
+func (r SynapsePrivateLinkHubResource) withUpdateFields(data acceptance.TestData) string {
+	template := r.template(data)
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_synapse_private_link_hub" "test" {
+  name                = "acctestsw%d"
```
Comment (manicminer) [inline azurerm/internal/services/synapse/synapse_private_link_hub_test.go:90]: As above
```suggestion
%[1]s

resource "azurerm_synapse_private_link_hub" "test" {
  name                = "acctestsw%[2]d"
```

File: website/docs/r/synapse_private_link_hub.html.markdown [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+# azurerm_synapse_private_link_hub
+
+Manages a Synapse Private Link Hub.
+
+## Example Usage
+
+```hcl
+resource "azurerm_synapse_private_link_hub" "example" {
+  name     = "example-resources"
+  location = "West Europe"
```
Comment (manicminer) [inline website/docs/r/synapse_private_link_hub.html.markdown:18]: It looks like `resource_group_name` is missing here

```suggestion
  name                = "example-resource"
  resource_group_name = "example-rg"
  location            = "West Europe"
```

File: website/docs/r/synapse_private_link_hub.html.markdown [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+}
+```
+
+## Arguments Reference
+
+The following arguments are supported:
+
+* `name` - (Required) The name which should be used for this Synapse Private Link Hub. Changing this forces a new Synapse Private Link Hub to be created.
+
```
Comment (manicminer) [inline website/docs/r/synapse_private_link_hub.html.markdown:27]: `resource_group_name` and `location` should be listed here

```suggestion

* `resource_group_name` - (Required) The name of the resource group in which to create the Synapse Private Link Hub. Changing this forces a new resource to be created.

* `location` - (Required) Specifies the Azure location where the Synapse Private Link Hub exists. Changing this forces a new resource to be created.

```

File: website/docs/r/synapse_private_link_hub.html.markdown [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+* `read` - (Defaults to 5 minutes) Used when retrieving the Synapse Private Link Hub.
+* `update` - (Defaults to 30 minutes) Used when updating the Synapse Private Link Hub.
+* `delete` - (Defaults to 30 minutes) Used when deleting the Synapse Private Link Hub.
+
+## Import
+
+Synapse Private Link Hub can be imported using the `resource id`, e.g.
+
+```shell
+terraform import azurerm_synapse_private_link_hub.example /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/group1/providers/Microsoft.Synapse/privateLinkHubs/privateLinkHub1
```
Comment (manicminer) [inline website/docs/r/synapse_private_link_hub.html.markdown:50]: We should double check that this format is correct once the id template is added to `resourceids.go`

**Review by manicminer — CHANGES_REQUESTED** [review]
Hi @owenfarrell, many thanks for this PR, this is off to a great start! I have made some suggestions inline below, most of which relate to some more recent changes/conventions within the provider so may not be immediately obvious when looking at some pre-existing resources for inspiration.

Did you generate the id parser/validator structs and functions? I noticed that the id template for this resource still needs to be added to `resourceids.go` (in the same directory as the resource and acceptance test files). Once it's added there, you can run `make generate` to (re)generate these - although given that the tests are passing there probably won't be any changes.

Your tests are great, I made a few small suggestions and we'll also need to add an import check test - this should create a resource, then try to create another identical one and verify that the import check in the Create function is working as expected.

If you can take a look at these, and all the tests continue to pass, this should be good to merge. Thanks again!

File: azurerm/internal/services/synapse/synapse_private_link_hub_test.go [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := client.Synapse.PrivateLinkHubsClient.Get(ctx, id.ResourceGroup, id.Name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			return utils.Bool(false), nil
+		}
+		return nil, fmt.Errorf("retrieving Synapse Private Link Hub %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
```
Comment (manicminer) [inline azurerm/internal/services/synapse/synapse_private_link_hub_test.go:65]:
```suggestion
		return nil, fmt.Errorf("retrieving %s: %+v", id, err)
```

**Comment by owenfarrell** [comment]
@manicminer Thanks for the review!

I folded in all of your proposed changes, updated `resources.go` to generate the parsing/validation code (my copypasta turned out to be accurate), and added an `_requiresImport` test.

Everything still checks out okay - let me know if there's anything else.

```
$ make acctests SERVICE='synapse' TESTARGS='-run=TestAccSynapsePrivateLinkHub_' TESTTIMEOUT='5m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./azurerm/internal/services/synapse -run=TestAccSynapsePrivateLinkHub_ -timeout 5m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccSynapsePrivateLinkHub_basic
=== PAUSE TestAccSynapsePrivateLinkHub_basic
=== RUN   TestAccSynapsePrivateLinkHub_requiresImport
=== PAUSE TestAccSynapsePrivateLinkHub_requiresImport
=== RUN   TestAccSynapsePrivateLinkHub_update
=== PAUSE TestAccSynapsePrivateLinkHub_update
=== CONT  TestAccSynapsePrivateLinkHub_basic
=== CONT  TestAccSynapsePrivateLinkHub_update
=== CONT  TestAccSynapsePrivateLinkHub_requiresImport
--- PASS: TestAccSynapsePrivateLinkHub_requiresImport (181.74s)
--- PASS: TestAccSynapsePrivateLinkHub_basic (184.69s)
--- PASS: TestAccSynapsePrivateLinkHub_update (210.74s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/synapse     213.359s
```

**Review by manicminer — APPROVED** [review]
Thanks @owenfarrell, this LGTM 🚀 

Screenshot 2021-07-15 at 14 51 38

**Pull request merged**